### ISSUE TITLE
weird miscompilation when linked to std dynamically on go 1.7.5

### ISSUE THREAD

**Issue by WGH-** [golang/go#19512]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
Tried 
* `go1.7.5 linux/amd64`
* `go1.7.5 linux/arm`
* `go1.8 linux/amd64` 
* `go1.8 linux/arm`

### What operating system and processor architecture are you using (`go env`)?
Correspondingly, 
```
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/tmp/gopath1.7.5"
GORACE=""
GOROOT="/tmp/go1.7.5"
GOTOOLDIR="/tmp/go1.7.5/pkg/tool/linux_amd64"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build493688201=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
```

```
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/tmp/gopath1.8"
GORACE=""
GOROOT="/tmp/go1.8"
GOTOOLDIR="/tmp/go1.8/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build192135816=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
PKG_CONFIG="pkg-config"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
```
```
GOARCH="arm"
GOBIN=""
GOEXE=""
GOHOSTARCH="arm"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/tmp/gopath1.7.5"
GORACE=""
GOROOT="/tmp/go1.7.5"
GOTOOLDIR="/tmp/go1.7.5/pkg/tool/linux_arm"
CC="gcc"
GOGCCFLAGS="-fPIC -marm -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build624821284=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
```

```
GOARCH="arm"
GOBIN=""
GOEXE=""
GOHOSTARCH="arm"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/tmp/gopath1.8"
GORACE=""
GOROOT="/tmp/go1.8"
GOTOOLDIR="/tmp/go1.8/pkg/tool/linux_arm"
GCCGO="gccgo"
GOARM=""
CC="gcc"
GOGCCFLAGS="-fPIC -marm -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build323933886=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
PKG_CONFIG="pkg-config"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
```

### What did you do?
First of all, I took precompiled go binaries from the official site and extracted the to /tmp to separate directories. And also prepared `GOROOT` and `GOPATH` correspondingly.

Then, I did `./bin/go install -buildmode=shared -linkshared  std`, which built a `std.so` that I was going to link to next. The build completed fine.

Next, I installed Scaleway CLI client using the provided command: `GO15VENDOREXPERIMENT=1 go get -u github.com/scaleway/scaleway-cli/cmd/scw`. The resulting binaries worked fine in all aforementioned environment. The binaries were linked statically so far. Just a test that static build works. Ok.

I tried building it dynamically: `GO15VENDOREXPERIMENT=1 go get -v -linkshared -u github.com/scaleway/scaleway-cli/cmd/scw`. The build itself finished successfully, But...

### What did you expect to see?
No problems. :)

### What did you see instead?
The binaries built by go1.7.5 didn't work. They just hang indefinitely on launch, without printing anything. `strace` doesn't reveal anything particularly interesting for me, they're no syscalls being made, so apparently the binaries hang in while true-like loop.

Binaries built by go1.8, on the other hand, work fine.

**Comment by WGH-** [comment]
This is actually not the first weird problem I encountered when linking dynamically on 1.7.5. My own small application occasionally crashes with `fatal error: concurrent map read and map write` very deep inside the standard lib.

```
fatal error: concurrent map read and map write
goroutine 539 [running]:
runtime.throw(0x7f1871c4acef, 0x21)
        /usr/lib/go/src/runtime/panic.go:566 +0x97 fp=0xc4204e5818 sp=0xc4204e57f8
runtime.mapaccess1_fast32(0x7f187209dda0, 0xc42009a000, 0xc40018c000, 0xc4204e58a0)
        /usr/lib/go/src/runtime/hashmap_fast.go:21 +0x1a6 fp=0xc4204e5840 sp=0xc4204e5818
runtime.(*_type).typeOff(0x7f18720ceae0, 0x7f180018c000, 0x7f18720e32e0)
        /usr/lib/go/src/runtime/type.go:227 +0x252 fp=0xc4204e58a0 sp=0xc4204e5840
runtime.additab(0x7f1872f2a0a8, 0x101)
        /usr/lib/go/src/runtime/iface.go:102 +0xee fp=0xc4204e5990 sp=0xc4204e58a0
runtime.getitab(0x7f18720ceae0, 0x7f18720e8ee0, 0xc4200a2001, 0xc4204e5a88)
        /usr/lib/go/src/runtime/iface.go:79 +0x1c0 fp=0xc4204e5a28 sp=0xc4204e5990
runtime.assertE2I2(0x7f18720ceae0, 0x7f18720e8ee0, 0xc4203cef00, 0xc4204e5ae8, 0x0)
        /usr/lib/go/src/runtime/iface.go:383 +0x80 fp=0xc4204e5a58 sp=0xc4204e5a28
fmt.(*pp).handleMethods(0xc42016c000, 0x76, 0xc4204e5c00)
        /usr/lib/go/src/fmt/print.go:554 +0xb6 fp=0xc4204e5b08 sp=0xc4204e5a58
fmt.(*pp).printArg(0xc42016c000, 0x7f18720e8ee0, 0xc4203cef00, 0x76)
        /usr/lib/go/src/fmt/print.go:665 +0x181 fp=0xc4204e5c00 sp=0xc4204e5b08
fmt.(*pp).doPrintf(0xc42016c000, 0x4a9b5f, 0x1c, 0xc4204e5e78, 0x1, 0x1)
        /usr/lib/go/src/fmt/print.go:985 +0x1267 fp=0xc4204e5ce8 sp=0xc4204e5c00
fmt.Sprintf(0x4a9b5f, 0x1c, 0xc4204e5e78, 0x1, 0x1, 0x7f1871b08661, 0x6cd980)
        /usr/lib/go/src/fmt/print.go:196 +0x6c fp=0xc4204e5d40 sp=0xc4204e5ce8
fmt.Errorf(0x4a9b5f, 0x1c, 0xc4204e5e78, 0x1, 0x1, 0xc4203cef00, 0x0)
        /usr/lib/go/src/fmt/print.go:205 +0x5c fp=0xc4204e5da0 sp=0xc4204e5d40
github.com/armon/go-socks5.(*Server).ServeConn(0xc420158030, 0x7f18724a4260, 0xc42000c028, 0x0, 0x0)
        /home/wgh/go/src/github.com/armon/go-socks5/socks5.go:163 +0x88c fp=0xc4204e5f78 sp=0xc4204e5da0
runtime.goexit()
        /usr/lib/go/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc4204e5f80 sp=0xc4204e5f78
created by github.com/armon/go-socks5.(*Server).Serve
        /home/wgh/go/src/github.com/armon/go-socks5/socks5.go:115 +0x78
```

At the some point I managed to compile an empty dynamically linked (that is, empty `main`) application that segfaults immediately on launch. I wasn't able to reproduce it (as in "compile another binary like this"), but I still have that crashing binary.

**Comment by bradfitz** [comment]
Closing, as it sounds like there's nothing to do if Go 1.8 works fine. (Just don't use Go 1.7.x)

**Comment by WGH-** [comment]
I've bisected this out of curiosity. It was actually fixed in 8b07ec20f7a971edce9eaf413b20e453c50858c6, making this issue a dup of #17594.

What really bugs me though it that even though the regression appeared in 1.7, and was fixed several months before 1.8 release, the fix was never backported to 1.7.

**Comment by ianlancetaylor** [comment]
Yes, that should probably have been backported to the 1.7 branch.  Sorry.
